tests/transports/smpp/test_client.py

```diff
@@ -0,0 +1,166 @@
+import logging
+from io import BytesIO
+
+from pytest import fixture, raises
+from smpp.pdu.operations import (
+    BindTransceiver,
+    BindTransceiverResp,
+    EnquireLink,
+    EnquireLinkResp,
+    Outbind,
+    SubmitSM,
+    SubmitSMResp,
+)
+from smpp.pdu.pdu_encoding import PDUEncoder
+from smpp.pdu.pdu_types import CommandStatus
+from trio.testing import memory_stream_pair
+
+from vumi2.transports.smpp.client import EsmeClient, EsmeResponseStatusError
+from vumi2.transports.smpp.smpp import SmppTransceiverTransportConfig
+
+
+def test_extract_pdu():
+    # Too short
+    assert EsmeClient.extract_pdu(bytearray.fromhex("00000004")) is None
+    # Not complete packet
+    assert (
+        EsmeClient.extract_pdu(bytearray.fromhex("00000020 050607080910111213141516"))
+        is None
+    )
+    # Complete packet
+    complete_packet = bytearray.fromhex("00000010 050607080910111213141516")
+    assert EsmeClient.extract_pdu(complete_packet) == bytearray.fromhex(
+        "00000010 050607080910111213141516"
+    )
+    assert complete_packet == b""
+    # More than one packet
+    packet_with_extra = bytearray.fromhex("00000010 050607080910111213141516 010203")
+    assert EsmeClient.extract_pdu(packet_with_extra) == bytearray.fromhex(
+        "00000010 050607080910111213141516"
+    )
+    assert packet_with_extra == bytearray.fromhex("010203")
+
+
+@fixture
+async def stream():
+    client, server = memory_stream_pair()
+    return (client, server)
+
+
+@fixture
+async def client_stream(stream):
+    return stream[0]
```

Comment:
Are `client_stream` and `server_stream` in the same call guaranteed to get each side of the same pair?

Reply:
If I'm understanding pytest fixtures correctly, then yes. 

The default fixture scope is `function`, so fixtures are generated once per function. So calling `client_stream`, will call `stream`, and return a response. Then calling `server_stream` will get the cached value for `stream`, and return a response. So they should both be from the same stream, at least on a per-test basis. Each test will get a new stream generated.

Reply:
Ah, of course. That means *all* fixtures will get the same `stream`, so if we want multiple streams we'd need multiple fixtures that make them.

Reply:
Yup. I've added a docstring there to explain that this is what we want, an easy way to share the same stream across fixtures and tests.